### ISSUE TITLE
[@typescript-eslint/no-unused-vars] False positive when parameter decorator has same name as parameter

### ISSUE THREAD

**Issue by ifeltsweet** [typescript-eslint/typescript-eslint#3874]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "plugins": ["@typescript-eslint"],
  "extends": ["plugin:@typescript-eslint/recommended"],
}
```

```TS
const command = (): ParameterDecorator => {
  return () => {
    //
  };
};

class Foo {
  bar(@command() command: string) {
    //
  }
}
```

**Expected Result**

No warning.

**Actual Result**

Warning on line 1:

```
'command' is assigned a value but never used.eslint@typescript-eslint/no-unused-vars.
```

**Additional Info**

Using different names for decorator and parameter does not produce the warning.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.31.0` |
| `@typescript-eslint/parser`        | `4.31.0` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.17.6` |

**Comment by Josh-Cena** [comment]
Should this be an upstream issue now? With decorators getting standardized, I think ESLint should eventually support this in the base rule. I don't know if it's worth investing the time to fix this in our extension.

**Comment by bradzacher** [comment]
@Josh-Cena no, we don't have an extension rule - we have a fork. We need to build our own support.

This is just a complicated thing to handle because of the way the scope analysis tooling works.

In a nutshell it works through a scope registering references and then it resolves them after it reaches the end of the scope.
Eg the reference created by the `@command` is resolved to the parameter, not the outer variable.

This is complicated to handle because technically it works this way due to how decorators are transpiled.
As an example [a default value for a parameter references the parameter, not the outer variable](https://typescript-eslint.io/play/#ts=4.9.5&sourceType=module&code=MYewdgzgLgBKC28CGYAmMC8MCMBuAUPgGYCuYwUAluDEQBQLJqZwiIqoCUMA3jAL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA).
